extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/OAuth2Template.java

```diff
@@ -0,0 +1,115 @@
+import org.wildfly.security.auth.realm.token.TokenValidator;
+import org.wildfly.security.auth.realm.token.validator.OAuth2IntrospectValidator;
+import org.wildfly.security.auth.server.SecurityDomain;
+import org.wildfly.security.auth.server.SecurityRealm;
+
+import io.quarkus.elytron.security.oauth2.runtime.oauth.ElytronOAuth2CallerPrincipal;
+import io.quarkus.elytron.security.oauth2.runtime.oauth.OAuth2AuthMechanismFactory;
+import io.quarkus.elytron.security.oauth2.runtime.oauth.OAuth2IdentityManager;
+import io.quarkus.elytron.security.runtime.*;
+import io.quarkus.runtime.RuntimeValue;
+import io.quarkus.runtime.annotations.Template;
+import io.undertow.security.idm.IdentityManager;
+import io.undertow.servlet.ServletExtension;
+import io.undertow.servlet.api.DeploymentInfo;
+
+/**
+ * The OAuth2 runtime security template class that provides methods for creating RuntimeValues for the deployment
+ * of OAuth2 security objects.
+ */
+@Template
+public class OAuth2Template {
+
+    /**
+     * Create the TokenSecurityRealm
+     *
+     * @return runtime wrapped TokenSecurityRealm
+     */
+    public RuntimeValue<SecurityRealm> createRealm(OAuth2Config config)
+            throws IOException, NoSuchAlgorithmException, CertificateException, KeyStoreException, KeyManagementException {
+        SSLContext sslContext = config.caCertFile.isPresent() ? createSLLContext(config) : SSLContext.getDefault();
+
+        TokenValidator validator = OAuth2IntrospectValidator.builder()
+                .clientId(config.clientId)
+                .clientSecret(config.clientSecret)
+                .tokenIntrospectionUrl(URI.create(config.introspectionUrl).toURL())
+                .useSslContext(sslContext)
+                .build();
+
+        TokenSecurityRealm tokenRealm = TokenSecurityRealm.builder()
+                .validator(validator)
+                .claimToPrincipal(claims -> new ElytronOAuth2CallerPrincipal(claims))
+                .build();
+
+        return new RuntimeValue<>(tokenRealm);
+    }
+
+    private SSLContext createSLLContext(OAuth2Config config)
+            throws IOException, CertificateException, NoSuchAlgorithmException, KeyStoreException, KeyManagementException {
+        InputStream is = new FileInputStream(config.caCertFile.get());
+
+        CertificateFactory cf = CertificateFactory.getInstance("X.509");
+        X509Certificate caCert = (X509Certificate) cf.generateCertificate(is);
+
+        TrustManagerFactory tmf = TrustManagerFactory
+                .getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
+        ks.load(null); // You don't need the KeyStore instance to come from a file.
+        ks.setCertificateEntry("caCert", caCert);
+
+        tmf.init(ks);
+
+        SSLContext sslContext = SSLContext.getInstance("TLS");
+        sslContext.init(null, tmf.getTrustManagers(), null);
+
+        return sslContext;
+    }
```

Comment:
Looking at this code make me think if JAX-RS client api (`guarkus-rest-client` will have it) can be used instead of using the wildfly component, it is really a form request which returns a simple JSON response, and the JAX-RS client API has SSLContext code in place as well... Only a suggestion that be discussed later if it would be of interest.